tmc-plugin-intellij/src/main/java/fi/helsinki/cs/tmc/intellij/ui/testresults/TestResultsPanel.java

```diff
@@ -1,111 +1,196 @@
+        gbc.insets.top = MARGIN;
+        this.resultsListConstraints = gbc;
+    }
+
+    private void createTestRows(List<TestResult> testResults) {
+        for (TestResult result : testResults) {
+            String details = getDetails(result);
+            TestResultRow row = createTestRow(result, details);
+            resultsList.add(row, resultsListConstraints);
+        }
+    }
+
+    private String getDetails(TestResult result) {
+        List<String> detailsRows = getDetailsOrMessageRows(result);
+        if (detailsRows == null) {
+            return null;
+        }
+        return String.join("\n", detailsRows);
+    }
+
+    private List<String> getDetailsOrMessageRows(TestResult res) {
+        return res.getDetailedMessage().size() > 0 ? res.getDetailedMessage() : res.getException();
+    }
+
+    private TestResultRow createTestRow(TestResult result, String details) {
+        if (result.isSuccessful()) {
+            return TestResultRow.createSuccessfulTestRow(
+                    result.getName(),
+                    result.getMessage(),
+                    details);
+        } else {
+            return TestResultRow.createFailedTestRow(
+                    result.getName(),
+                    result.getMessage(),
+                    details);
+        }
+    }
+
+    private void createValidationRows(ValidationResult validationResult) {
+        Map<File, List<ValidationError>> failedFiles = validationResult.getValidationErrors();
+
+        for (Map.Entry<File, List<ValidationError>> failedFile : failedFiles.entrySet()) {
+            String filename = failedFile.getKey().toString();
+            List<ValidationError> errors = failedFile.getValue();
+            createValidationRowForEachError(filename, errors);
         }
+    }
+
+    private void createValidationRowForEachError(String filename, List<ValidationError> errors) {
 
-        bar.setMinimum(0);
-        bar.setMaximum(100);
-        bar.setBorderPainted(false);
-        bar.setStringPainted(true);
-        bar.setValue((int) (100 * ((double) success / results.size())));
-        basePanel.repaint();
+        for (ValidationError error : errors) {
+            String message = String.format(
+                    "Line %d:    %s",
+                    error.getLine(),
+                    error.getMessage());
+
+            TestResultRow row =
+                    TestResultRow.createValidationRow(filename, message, error.getSourceName());
+            resultsList.add(row, resultsListConstraints);
+        }
     }
 
-    private Color getColor(boolean successful) {
-        if (successful) {
-            return Color.GREEN;
+    private double testPassRatio(List<TestResult> tests) {
+        int total = tests.size();
+        int passed = 0;
+
+        for (TestResult test : tests) {
+            if (test.isSuccessful()) {
+                passed++;
+            }
+        }
+
+        if (total == 0) {
+            return 0;
```

Comment:
Good point. I recall choosing 0 because of #113's effects. There's currently no checks for if the tests fail to compile (e.g. `mvn test-compile` fails because javac isn't found from `PATH`), causing `tests.size()` to be 0. In this case, no tests were run so it passed 0% of the tests instead of 100% of the tests.

But yeah, you're right - if there really were no tests, it should be 1. Shouldn't set it to 0 as a workaround for a bug. I'll change it.